child_process: validate fork/execFile arguments

Trott opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Bug fixes and new features should include tests and possibly a benchmark.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX) or `vcbuild test nosign` (Windows) passes
- [x] a test and/or benchmark is included
- [x] the commit message follows commit guidelines


##### Affected core subsystem(s)
<!-- provide affected core subsystem(s) (like doc, cluster, crypto, etc) -->
child_process

##### Description of change
<!-- provide a description of the change below this comment -->

I took the tests that @ChuckLangford wrote in https://github.com/nodejs/node/pull/4508 and implemented code such that they pass. This fixes https://github.com/nodejs/node/issues/2681.

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/3081/

Trott commented:
CI is green. /cc @jasnell @bnoordhuis @cjihrig

bnoordhuis commented on lib/child_process.js:
```diff
@@ -19,15 +19,20 @@ const ChildProcess = exports.ChildProcess = child_process.ChildProcess;
+  var options = {};
+  var args = [];
+  var pos = 1;
+  if (Array.isArray(arguments[pos])) {
```
Can you check that `pos < arguments.length`?  I believe V8 still deoptimizes on out-of-bounds arguments access.

Trott replied:
Happy to make the change. A couple of questions, though, mostly to make sure I'm understanding things correctly:

* The code I'm replacing accessed `arguments[1]` without any bounds check, so it was already not being optimized. Or was there something else going on there that made that code optimizable?

* This isn't going to be something that's benchmark-able because each `fork()` call is going to spin up a completely separate Node.js instance and any performance improvement in the single call to `fork()` is going to be lost in the overhead of launching the separate instance. Or am I wrong and this may be reasonably benchmark-able perhaps using some trick I'm not thinking of?

bnoordhuis replied:
> The code I'm replacing accessed arguments[1] without any bounds check, so it was already not being optimized.

Yes, that's quite possibly the case.

> This isn't going to be something that's benchmark-able because each fork() call is going to spin up a completely separate Node.js instance

That's right but it would be good for consistency.  Elsewhere in lib/child_process.js we're quite diligent about checking `arguments.length`.  Also, it makes `--trace_deopt` less noisy, which is never a bad thing.

Trott replied:
OK, added in the checks for `pos < arguments.length`.

bnoordhuis commented:
Mostly LGTM, I think.

jasnell commented:
LGTM

Trott commented:
Addressed optimization nits from @bnoordhuis, rebased, force pushed.

CI again: https://ci.nodejs.org/job/node-test-pull-request/3099/

And, because this is `semver-major`, adding the `ctc-agenda` label.

bnoordhuis commented:
d07430aef7a6fd230768bff8adfdd22cafd6220f LGTM.

Trott commented:
SInce it's semver-major: @nodejs/ctc (Will be on agenda for tomorrow's meeting.)

rvagg commented:
no objections from today's CTC meeting for this

Trott commented:
Landed in 0548e5d12ac32b6d58c8aac8bcf4dc8cff708b1f and 99cfd53097ef646e0e61a0690c9b80ff9ae80d68. Thanks, @ChuckLangford!

ChuckLangford commented:
This has been a fantastic learning experience for me. Thanks @Trott
